Use the updated SDK

emirotin opened the pull request:
Completely untested, just contains the general idea.

NB: there was a JS file `actions/enter.js` that didn't have the corresponding coffee source file.

emirotin commented:
Ping @pimterry this is the PR for the CLI and it's apparently in a good shape.

pimterry commented:
I've done some basic testing, and a few small fixes for things have fallen out, which I've pushed, but otherwise it looks good at a first look to me.

@jviotti Eugene mentioned before that it was worth talking to you about testing. Do we have a thorough manual testing process for this? What do we need to consider this mergeable?

@emirotin There's another few related repos that depend on `resin-sdk` which aren't updated yet (resin-cli-auth, resin-cli-events, resin-config-json, resin-device-config, resin-device-init, resin-image-manager). I'm going to quickly go around those and put in PRs to update them all similarly in just a minute, let me know if that's not worthwhile.

Also, `resin-sync` is pretty out of date here too, missing quite a very major bumps (2.2.0 vs 6.0.1). Worth bumping that up here too?

emirotin commented:
All the question re the CLI-related repos are better to address to Juan.
For resin-sync talk to Kostas too.

On Wed, Jan 11, 2017 at 6:07 PM, Tim Perry <notifications@github.com> wrote:

> I've done some basic testing, and a few small fixes for things have fallen
> out, which I've pushed, but otherwise it looks good at a first look to me.
>
> @jviotti <https://github.com/jviotti> Eugene mentioned before that it was
> worth talking to you about testing. Do we have a thorough manual testing
> process for this? What do we need to consider this mergeable?
>
> @emirotin <https://github.com/emirotin> There's another few related repos
> that depend on resin-sdk which aren't updated yet (resin-cli-auth,
> resin-cli-events, resin-config-json, resin-device-config,
> resin-device-init, resin-image-manager). I'm going to quickly go around
> those and put in PRs to update them all similarly in just a minute, let me
> know if that's not worthwhile.
>
> Also, resin-sync is pretty out of date here too, missing quite a very
> major bumps (2.2.0 vs 6.0.1). Worth bumping that up here too?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/resin-io/resin-cli/pull/413#issuecomment-271892398>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAgGCEJhiyxlD89ChfpZH_oMCGN3G-9Rks5rRPAwgaJpZM4LPPv9>
> .
>



-- 
Eugene Mirotin
Senior Frontend Engineer
site: Resin.io <https://resin.io/>, twitter: @resin_io
<https://twitter.com/resin_io>

pimterry commented:
@jviotti I had a look around the other cli modules. I can go round and update resin-sdk in them, but in each case it's going to be a debatable and breaking change, since they'll need to get either resin-sdk's options or a pre-instantiated resin-sdk passed in - they can't just blindly use the module any more.

An alternative option is that we could create a separate `resin-sdk-preconfigured` module, with a preinstantiated ready-to-use singleton SDK, with a standard config for the CLI (i.e. [resin-sdk.coffee](https://github.com/resin-io/resin-cli/pull/413/files#diff-e5b38a347cc33361310a331ba16e8fbc) from this PR, as a standalone module), and then swap their dependencies all over to that, to immediately solve this. I'd then integrate that into this PR too.

Thoughts? (Cc @emirotin, @Page- too). The alternative is that we just ignore this completely, until there's something we need from the SDK in these modules, but that feels a bit untidy, and this will definitely come round to bite us eventually.

---

@lekkas See my previous comment - I'm going around updating dependencies, and I noticed resin-cli is still on v2.2.0 of resin-sync (vs 6.0.1 now). A quick skim over there suggests the changes since then would be easy to integrate here, but might have some breaking behaviour for end users. Is it worth pulling that in here anyway, or not yet?

jviotti commented:
> An alternative option is that we could create a separate resin-sdk-preconfigured module, with a preinstantiated ready-to-use singleton SDK, with a standard config for the CLI (i.e. resin-sdk.coffee from this PR, as a standalone module), and then swap their dependencies all over to that, to immediately solve this. I'd then integrate that into this PR too.

> Thoughts? (Cc @emirotin, @Page- too). The alternative is that we just ignore this completely, until there's something we need from the SDK in these modules, but that feels a bit untidy, and this will definitely come round to bite us eventually.

Sounds good to me. Alternatively we could refactor all the modules to accept an SDK instance on instantiation, and then we would only list the Resin SDK as a dependency of the main Resin CLI module, and broadcast everywhere, but that sounds like more work.

emirotin commented:
If the modules are only used by the CLI either way works for me - either DI
(consistent with what we do in other places) or a preconfigured singleton
(I would then notice this fact in the module name by either mentioning the
CLI or Node).
If some of them may be of any use in the browser the DI is obviously better.

On Wed, Jan 11, 2017 at 11:04 PM, Juan Cruz Viotti <notifications@github.com
> wrote:

> An alternative option is that we could create a separate
> resin-sdk-preconfigured module, with a preinstantiated ready-to-use
> singleton SDK, with a standard config for the CLI (i.e. resin-sdk.coffee
> from this PR, as a standalone module), and then swap their dependencies all
> over to that, to immediately solve this. I'd then integrate that into this
> PR too.
>
> Thoughts? (Cc @emirotin <https://github.com/emirotin>, @Page-
> <https://github.com/Page-> too). The alternative is that we just ignore
> this completely, until there's something we need from the SDK in these
> modules, but that feels a bit untidy, and this will definitely come round
> to bite us eventually.
>
> Sounds good to me. Alternatively we could refactor all the modules to
> accept an SDK instance on instantiation, and then we would only list the
> Resin SDK as a dependency of the main Resin CLI module, and broadcast
> everywhere, but that sounds like more work.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/resin-io/resin-cli/pull/413#issuecomment-271978589>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAgGCFlEPNNHalYOZrn676u7hzZsaBNeks5rRTXIgaJpZM4LPPv9>
> .
>



-- 
Eugene Mirotin
Senior Frontend Engineer
site: Resin.io <https://resin.io/>, twitter: @resin_io
<https://twitter.com/resin_io>

pimterry commented:
I was pretty convinced by all the above as a strong argument for the injection approach, and I got a reasonable distance into that, but there's some big downsides, and I'm going for the standalone module option in the end. A few reasons, for reference:

1. Injection is quite a bit more complex that it looks, because it's not just adding injection to the interface of all those modules, but also typically injection between every single file within them, to get the injected resinSdk from the entry point to every other part of the code.
2. Every test suite ends up duplicating the exact same setup and injection for the same preconfigured SDK.
3. You can get the benefits in other ways. For browser/node versions, for example, we can use `main` and `browser` in package.json to provide _both_ preconfigured in one module, and automatically use the right one. We can also sneakily include a SDK preconfigured for tests as `resin-preconfigured-sdk/tests/test-sdk` too and share that everywhere too, as a nice bonus.
4. It's a first step that still leaves us open to moving to injection later if we'd like, without doing all the effort now.

I'm currently putting together the preconfigured module at https://github.com/resin-io-modules/resin-sdk-preconfigured and testing it in a few of these modules. I'm doing this in TypeScript, just using Page's setup on [typed-error](https://github.com/resin-io/typed-error/tree/typescript), since I think that's the way we're aiming now? Shout if you're not happy moving in that direction

@emirotin For now I haven't included Node or cli in the name (as in your comment). As in point 3 above, I don't think this is necessarily specific to either Node or the cli. Let me know if you disagree strongly though - it's easy to change up until we publish it.

I'll open a review over there shortly, once I've got a version ready.

emirotin commented:
Ah good point about the main/browser thing, missed that yesterday.
Sounds good to me.

On Thu, Jan 12, 2017 at 5:07 PM, Tim Perry <notifications@github.com> wrote:

> I was pretty convinced by all the above as a strong argument for the
> injection approach, and I got a reasonable distance into that, but there's
> some big downsides, and I'm going for the standalone module option in the
> end. A few reasons, for reference:
>
>    1. Injection is quite a bit more complex that it looks, because it's
>    not just adding injection to the interface of all those modules, but also
>    typically injection between every single file within them, to get the
>    injected resinSdk from the entry point to every other part of the code.
>    2. Every test suite ends up duplicating the exact same setup and
>    injection for the same preconfigured SDK.
>    3. You can get the benefits in other ways. For browser/node versions,
>    for example, we can use main and browser in package.json to provide
>    *both* preconfigured in one module, and automatically use the right
>    one. We can also sneakily include a SDK preconfigured for tests as
>    resin-preconfigured-sdk/tests/test-sdk too and share that everywhere
>    too, as a nice bonus.
>    4. It's a first step that still leaves us open to moving to injection
>    later if we'd like, without doing all the effort now.
>
> I'm currently putting together the preconfigured module at
> https://github.com/resin-io-modules/resin-sdk-preconfigured and testing
> it in a few of these modules. I'm doing this in TypeScript, just using
> Page's setup on typed-error
> <https://github.com/resin-io/typed-error/tree/typescript>, since I think
> that's the way we're aiming now? Shout if you're not happy moving in that
> direction
>
> @emirotin <https://github.com/emirotin> For now I haven't included Node
> or cli in the name (as in your comment). As in point 3 above, I don't think
> this is necessarily specific to either Node or the cli. Let me know if you
> disagree strongly though - it's easy to change up until we publish it.
>
> I'll open a review over there shortly, once I've got a version ready.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/resin-io/resin-cli/pull/413#issuecomment-272171522>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAgGCE5j28NtOamnZaE7Ehz8I_ovW-IDks5rRjOxgaJpZM4LPPv9>
> .
>



-- 
Eugene Mirotin
Senior Frontend Engineer
site: Resin.io <https://resin.io/>, twitter: @resin_io
<https://twitter.com/resin_io>

pimterry commented:
This is now rebased on top of recent master changes, and I've updated this codebase and each of the other relevant deps to `resin-sdk-preconfigured`, and done releases for all of those (except resin-sync, since we're way behind on versions here at the moment anyway, and I didn't hear anything from @lekkas).

I've done a tiny bit of quick local testing with this too, logging in, listing devices and apps, and creating a new app. @jviotti Anything specific I need to test on this before it goes out?

Anything else I'm missing? If not, if people could approve once they're happy that'd be great.

emirotin commented:
I think one thing that needs testing is provisioning the device that
involves fetching and patching the OS image.
Quickstart (which is the full process) is another thing to check quickly.

On Wed, Jan 25, 2017 at 8:47 PM, Tim Perry <notifications@github.com> wrote:

> This is now rebased on top of recent master changes, and I've updated this
> codebase and each of the other relevant deps to resin-sdk-preconfigured,
> and done releases for all of those (except resin-sync, since we're way
> behind on versions here at the moment anyway, and I didn't hear anything
> from @lekkas <https://github.com/lekkas>).
>
> I've done a tiny bit of quick local testing with this too, logging in,
> listing devices and apps, and creating a new app. @jviotti
> <https://github.com/jviotti> Anything specific I need to test on this
> before it goes out?
>
> Anything else I'm missing? If not, if people could approve once they're
> happy that'd be great.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/resin-io/resin-cli/pull/413#issuecomment-275196357>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAgGCAQuD5Yy_3L_kU5zJGBge8Jz-0hLks5rV5jKgaJpZM4LPPv9>
> .
>



-- 
Eugene Mirotin
Senior Frontend Engineer
site: Resin.io <https://resin.io/>, twitter: @resin_io
<https://twitter.com/resin_io>

lekkas commented:
@pimterry `resin-sync@6.0.1` had breaking changes, so it'd be best to stick with its current version for now - my plan is to update it in https://github.com/resin-io/resin-cli/pull/416 if it's fine with you guys.

In a relevant note, thank you for suggesting `Octobox` today, you saved the day wrt ping notifications :)

pimterry commented:
Thanks @lekkas, I'll leave that to you then.

---

@emirotin I've tested login -> create app -> quickstart -> happily online device. That all seems to work fine.

I'm not sure exactly what you meant by "provisioning the device that involves fetching and patching the OS image". I've tested:
* `resin os download raspberrypi3 -o rpi3.img`
* `resin os configure rpi3.img <uuid>`
* `resin os initialize rpi3.img --type 'raspberrypi3'`
* -> Device comes online

Does that cover it?

emirotin commented:
Yeah I think that's it. Streaming OS download was my biggest concern as
it's the code I made the biggest changes to (because of transition from
request to fetch). Though it's covered by the integration tests.

On Thu, Jan 26, 2017 at 10:45 PM, Tim Perry <notifications@github.com>
wrote:

> Thanks @lekkas <https://github.com/lekkas>, I'll leave that to you then.
> ------------------------------
>
> @emirotin <https://github.com/emirotin> I've tested login -> create app
> -> quickstart -> happily online device. That all seems to work fine.
>
> I'm not sure exactly what you meant by "provisioning the device that
> involves fetching and patching the OS image". I've tested:
>
>    - resin os download raspberrypi3 -o rpi3.img
>    - resin os configure rpi3.img <uuid>
>    - resin os initialize rpi3.img --type 'raspberrypi3'
>    - -> Device comes online
>
> Does that cover it?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/resin-io/resin-cli/pull/413#issuecomment-275492392>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAgGCHi516IkUOVIKytldqvfBmh-mwjnks5rWPfXgaJpZM4LPPv9>
> .
>



-- 
Eugene Mirotin
Senior Frontend Engineer
site: Resin.io <https://resin.io/>, twitter: @resin_io
<https://twitter.com/resin_io>
